### ISSUE TITLE
Support Option-wrapped lists

### ISSUE THREAD

**Issue by cmeeren** [ChilliCream/graphql-platform#7472]
### Product

Hot Chocolate

### Is your feature request related to a problem?

Currently working on #6884 by adapting the type interceptor from [this gist](https://gist.github.com/Stock44/0f465a56fba5095fbf078b1d0ee4526a). Unfortunately, I seem to be blocked by HotChocolate eagerly prohibiting Option-wrapped lists without actually calling the built-in [OptionTypeConverter](https://github.com/ChilliCream/graphql-platform/blob/92c99dffa5be4b1e70bae442a10ca5ebd8876de0/src/HotChocolate/Core/src/Types.FSharp/OptionTypeConverter.fs). That converter converts all option-wrapped values to their underlying representations (or `null` if the option value is `None`), and vice versa. However, HC does not invoke the converter and instead returns `EXEC_LIST_TYPE_NOT_SUPPORTED` whenever I try to query an option-wrapped list (array, in my case, though I believe it applies to all enumerables).

### The solution you'd like

For example, make HotChocolate run any converters before determining whether the type is a supported list type.

**Comment by cmeeren** [comment]
For output fields, this can be solved with field middleware. However, for input types and arguments, I do not see the ability to use middleware.

**Comment by michaelstaib** [comment]
field middleware is costly performance wise only use when there is async code on that same field. Never use it to just convert a thing...

**Comment by michaelstaib** [comment]
It might be that result formatters and input formatters are enough.

**Comment by cmeeren** [comment]
Thanks, I'll get a project with unit tests running and let you know.

(By the way, adding the middleware to the field definition seems to make this work for input types and arguments, too. No idea why.)

**Comment by cmeeren** [comment]
As mentioned in https://github.com/ChilliCream/graphql-platform/issues/6884#issuecomment-2363787392, a repo is up at [cmeeren/HotChocolate.FSharp](https://github.com/cmeeren/HotChocolate.FSharp).

The important code is in `Library.fs` in the `HotChocolate.FSharp` project.

The part about unwrapping top-level option-wrapped lists is [here](https://github.com/cmeeren/HotChocolate.FSharp/blob/8adb63378f8c6e70a76534e8816b32b654ab4a99/HotChocolate.FSharp/Library.fs#L208-L217).

Considering that that code may be merged into this repo at some point for proper built-in F# support in HC, you are more than welcome to look at the rest of the code in that file, specifically the `Helpers` module (and the short `FSharpNullabilityInterceptor` type definition). Additionally, for the same reason, I'd really appreciate some input on how to support `Option`-wrapped lists inside lists, and `Option`-wrapped enums inside lists.

**Comment by michaelstaib** [comment]
Yeah middleware is not meant for this ... try using a formatter .... `Definition.FormatterDefinitions` formatters work almost the same but are sync

**Comment by michaelstaib** [comment]
```
definition.FormatterDefinitions.Add(
            new ResultFormatterDefinition((ctx, result) => unwrap(result)));
```

**Comment by michaelstaib** [comment]
each input field btw can have Input formatters ... they exist on both sides ....

**Comment by cmeeren** [comment]
Excellent, thank you! I'll have a look at this.

If you have any insight about the other two items (`Option`-wrapped lists/enums/possibly other cases? and `Async<_>` support), then I think we'll be in a good place very soon.

**Comment by cmeeren** [comment]
I can confirm that worked great, thanks!

Did not need them for inputs. Again I have no idea why it works without it, but I have bigger fish to fry, so I'll leave that.

**Comment by cmeeren** [comment]
Just want to state that [FSharp.HotChocolate](https://github.com/cmeeren/FSharp.HotChocolate), now available on NuGet, has support for this.

@michaelstaib, I'll leave to you the decision of whether to treat that as sufficient to close this issue, or wait until support is added directly to HotChocolate.

**Comment by michaelstaib** [comment]
Can we pull the FSharp support entirely and your package deals with this?

**Comment by michaelstaib** [comment]
cc @glen-84

**Comment by cmeeren** [comment]
I guess that depends on what the F# support is? I know about the Option type converter, but not more.

Also, it as mentioned in some other issues, it may require some changes/additions to extension points the library can hook into.

**Comment by cmeeren** [comment]
FSharp.HotChocolate now includes `OptionTypeFormatter` and no longer has a dependency on HotChocolate.Types.FSharp. So if you want, HotChocolate.Types.FSharp can be deprecated with a note to use FSharp.HotChocolate instead

By the way, I think it would be useful to have a doc section or similar about F# where you point users to FSharp.HotChocolate.

**Comment by michaelstaib** [comment]
OK ... we will remove F# support with V15 and leave it to you.

**Comment by michaelstaib** [comment]
I however do not think that more extension points are needed ... I think there might only be knowledge transfer needed ... but we can discuss it.

**Comment by cmeeren** [comment]
Absolutely, if there are existing ways do accomplish [what we need](https://github.com/cmeeren/FSharp.HotChocolate/issues), I'm all ears.

**Comment by cmeeren** [comment]
> I however do not think that more extension points are needed ... I think there might only be knowledge transfer needed ... but we can discuss it.

@michaelstaib, about that knowledge transfer and discussion:

Since HC 15 removes F# support, I think FSharp.HotChocolate should aim for a feature complete 1.x release by the time HC 15 is out.

I have created a [1.0 milestone](https://github.com/cmeeren/FSharp.HotChocolate/milestone/1) that lists the issues where I need this "knowledge transfer" or more extension points. I hope the issue descriptions (or the HC issues/discussions linked to there) are sufficient. Please don't hesitate to let me know if you need more information.

**Comment by michaelstaib** [comment]
Yes, we have this on our list. However, the next 2 month we will be focused on fusion query planning.

**Comment by michaelstaib** [comment]
> https://github.com/ChilliCream/graphql-platform/releases/tag/16.0.0-p.8.2

can you check if this works for you? We can cherry pick it once we know it solves your issues

**Comment by cmeeren** [comment]
After updating from 15 to 16, I am encountering lots of compilation errors related to my type interceptors. The signature of `OnAfterInitialize` has changed. I have no idea what the migration path is. Here is my current code - what should I do instead?

```f#
    override this.OnAfterInitialize(discoveryContext, definition) =
        match definition with
        | :? ObjectTypeDefinition as objectDef -> objectDef.Fields |> Seq.iter addUnwrapUnionFormatter
        | _ -> ()
```

**Comment by michaelstaib** [comment]
ObjectTypeDefinition => ObjectTypeConfiguration
DefinitionBase = TypeSystemConfiguration

**Comment by michaelstaib** [comment]
Can you ping me on slack then we can iterate a bit quicker on this.

**Comment by cmeeren** [comment]
Got it now, thanks. Again though, as mentioned in https://github.com/ChilliCream/graphql-platform/pull/8722#issuecomment-3307492229, `ExtendedType` is internal, so I can't access the new `RegisterNonEssentialWrapperTypes`.

**Comment by glen-84** [comment]
@cmeeren What is the status here?

**Comment by cmeeren** [comment]
If you are talking about the original issue, then that was fixed long ago: https://github.com/ChilliCream/graphql-platform/issues/7472#issuecomment-2364709481

The conversion has since deviated from the original topic, but even that has now been resolved.

In other words, I believe this issue can be closed.
